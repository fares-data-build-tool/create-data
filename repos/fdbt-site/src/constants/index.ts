export const OPERATOR_COOKIE: string = 'fdbt-operator-cookie';

export const SERVICE_COOKIE: string = 'fdbt-service-cookie';

export const FARETYPE_COOKIE: string = 'fdbt-faretype-cookie';

export const FEEDBACK_LINK: string = 'mailto:fdbt@transportforthenorth.com';

export const GOVUK_LINK: string = 'https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/';

export const CSV_INPUT_METHOD_COOKIE: string = 'fdbt-csv-input-method-cookie';

export const MANUAL_INPUT_METHOD_COOKIE: string = 'fdbt-manual-input-method-cookie';

export const INT_MAP_INPUT_METHOD_COOKIE: string = 'fdbt-interactive-map-input-method-cookie';
