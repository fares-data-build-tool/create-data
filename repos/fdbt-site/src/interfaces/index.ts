import { NextApiRequest, NextPageContext } from 'next';
import { DocumentContext } from 'next/document';
import { IncomingMessage, ServerResponse } from 'http';
import { ReactElement } from 'react';

// Session Attributes and Cookies

export interface Session {
    session: Express.Session;
}

export type NextApiRequestWithSession = NextApiRequest & Session;

export type NextPageContextWithSession = NextPageContext & {
    req: Session;
};

export type DocumentContextWithSession = DocumentContext & {
    req: Session;
};

export type IncomingMessageWithSession = IncomingMessage & Session;

export interface CookiePolicy {
    essential: boolean;
    usage: boolean;
}

export interface ProductInfo {
    productName: string;
    productPrice: string;
}

export interface ProductInfoWithErrors extends ProductInfo {
    errors: ErrorInfo[];
}

export interface DurationValidInfo {
    amount: string;
    duration: string;
    errors: ErrorInfo[];
}

export interface InputMethodInfo {
    inputMethod: string;
}

export interface Journey extends JourneyWithErrors {
    directionJourneyPattern?: string;
    inboundJourney?: string;
    outboundJourney?: string;
}

export interface JourneyWithErrors {
    errors?: ErrorInfo[];
}

export interface TicketRepresentationAttribute {
    name: string;
}

export interface TicketRepresentationAttributeWithErrors {
    errors: ErrorInfo[];
}

export interface Operator {
    name: string;
    nocCode: string;
}

export interface OperatorGroup {
    name: string;
    operators: Operator[];
}

export interface MultiOperatorInfo {
    nocCode: string;
    services: string[];
}

export interface TermTimeAttribute {
    termTime: boolean;
}

export type WithErrors<T> = {
    errors: ErrorInfo[];
} & T;

export interface UserDataUploadsProps {
    csvUploadApiRoute: string;
    csvUploadHintText: string;
    csvUploadTitle: string;
    guidanceDocDisplayName: string;
    guidanceDocAttachmentUrl: string;
    guidanceDocSize: string;
    csvTemplateDisplayName: string;
    csvTemplateAttachmentUrl: string;
    csvTemplateSize: string;
    errors: ErrorInfo[];
    detailSummary?: string;
    detailBody?: ReactElement;
    showPriceOption?: boolean;
    poundsOrPence?: string | null;
    csrfToken: string;
}

export interface GroupDefinitionWithErrors extends GroupDefinition {
    errors: ErrorInfo[];
}

export interface NumberOfStagesAttributeWithError {
    errors: ErrorInfo[];
}

export interface FareStagesAttribute {
    fareStages: string;
}

export interface FareStagesAttributeWithErrors {
    errors: ErrorInfo[];
}

export interface CsvUploadAttributeWithErrors {
    errors: ErrorInfo[];
    poundsOrPence?: string;
}

export interface GroupTicketAttribute {
    maxGroupSize: string;
}

export interface GroupTicketAttributeWithErrors extends GroupTicketAttribute {
    errors: ErrorInfo[];
}

export interface NumberOfProductsAttribute {
    numberOfProductsInput: string;
}

export interface NumberOfProductsAttributeWithErrors {
    errors: ErrorInfo[];
}

export interface MultipleProductAttribute {
    products: MultiProduct[];
}

export interface MultipleProductAttributeWithErrors extends MultipleProductAttribute {
    errors: ErrorInfo[];
}

export interface SchoolFareTypeAttribute {
    schoolFareType: string;
}

export interface MultipleOperatorsAttribute {
    selectedOperators: Operator[];
}

export interface MultipleOperatorsAttributeWithErrors extends MultipleOperatorsAttribute {
    errors: ErrorInfo[];
}

export interface ServiceListAttribute {
    selectedServices: string[];
}

export interface ServiceListAttributeWithErrors {
    errors: ErrorInfo[];
}

// Miscellaneous

export interface PassengerAttributes {
    passengerTypeDisplay: string;
    passengerTypeValue: string;
}

export interface BaseReactElement {
    id: string;
    name: string;
    label: string;
    defaultValue?: string;
    defaultChecked?: boolean;
    options?: string[];
}

export interface InputCheck {
    error: string;
    input: string;
    id: string;
}

export interface ErrorInfo {
    errorMessage: string;
    id: string;
    userInput?: string;
}

export interface Feedback {
    question: string;
    answer: string;
}

export interface PriceEntryError {
    input: string;
    locator: string;
}

export interface ResponseWithLocals extends ServerResponse {
    locals: {
        nonce: string;
        csrfToken: string;
    };
}

export interface PremadeTimeRestriction {
    name: string;
    contents: FullTimeRestriction[];
}

// AWS and Reference Data (e.g. NOC, TNDS, NaPTAN datasets)

export interface Stop {
    stopName: string;
    naptanCode: string;
    atcoCode: string;
    localityCode: string;
    localityName: string;
    parentLocalityName: string;
    qualifierName?: string;
    indicator?: string;
    street?: string;
}

export interface StopPoint {
    stopPointRef: string;
    commonName: string;
}

export interface S3NetexFile {
    name: string;
    noc: string | undefined;
    reference: string;
    fareType: string;
    productNames?: string;
    passengerType: string;
    serviceNames?: string;
    lineName?: string;
    zoneName?: string;
    sopNames: string;
    date: string;
    signedUrl: string;
    fileSize: number;
}

/* eslint-disable camelcase */
export interface CognitoIdToken {
    sub: string;
    aud: string;
    email_verified: boolean;
    event_id: string;
    'custom:noc': string;
    token_use: string;
    auth_time: number;
    iss: string;
    'cognito:username': string;
    exp: number;
    iat: number;
    email: string;
    'custom:contactable': string;
    'custom:schemeOperator': string;
    'custom:schemeRegionCode': string;
}

// Ticket Types

export type Ticket =
    | PointToPointTicket
    | GeoZoneTicket
    | PeriodMultipleServicesTicket
    | FlatFareTicket
    | SchemeOperatorTicket;

export type PointToPointTicket = SingleTicket | ReturnTicket;

export interface BaseTicket {
    nocCode: string;
    type: string;
    passengerType: string;
    ageRange?: string;
    ageRangeMin?: string;
    ageRangeMax?: string;
    proof?: string;
    proofDocuments?: string[];
    email: string;
    uuid: string;
    timeRestriction: FullTimeRestriction[];
    ticketPeriod: TicketPeriod;
}

export interface BasePointToPointTicket extends BaseTicket {
    operatorShortName: string;
    lineName: string;
    serviceDescription: string;
    products: BaseProduct[];
}

export interface SingleTicket extends BasePointToPointTicket {
    fareZones: FareZone[];
    termTime: boolean;
}

export interface ReturnTicket extends BasePointToPointTicket {
    inboundFareZones: FareZone[];
    outboundFareZones: FareZone[];
    returnPeriodValidity?: ReturnPeriodValidity;
}

export type PeriodTicket = PeriodGeoZoneTicket | PeriodMultipleServicesTicket;

export interface BasePeriodTicket extends BaseTicket {
    operatorName: string;
    products: ProductDetails[];
}

export interface PeriodGeoZoneTicket extends BasePeriodTicket {
    zoneName: string;
    stops: Stop[];
}

export interface Product {
    productName: string;
    productPrice: string;
    productDuration?: string;
    productValidity?: string;
    productDurationUnits?: string;
    productEndTime?: string;
}

export interface MultiOperatorGeoZoneTicket extends PeriodGeoZoneTicket {
    additionalNocs: string[];
}

export type GeoZoneTicket = PeriodGeoZoneTicket | MultiOperatorGeoZoneTicket;

export interface PeriodMultipleServicesTicket extends BasePeriodTicket {
    selectedServices: SelectedService[];
    termTime: boolean;
}

export interface MultiOperatorMultipleServicesTicket extends PeriodMultipleServicesTicket {
    additionalOperators: {
        nocCode: string;
        selectedServices: SelectedService[];
    }[];
}

export interface FlatFareTicket extends BaseTicket {
    operatorName: string;
    products: FlatFareProductDetails[];
    selectedServices: SelectedService[];
    termTime: boolean;
}

export interface BaseGroupTicket {
    nocCode: string;
    type: string;
    groupDefinition: GroupDefinition;
    email: string;
    uuid: string;
}

export interface SchemeOperatorTicket {
    schemeOperatorName: string;
    schemeOperatorRegionCode: string;
    nocCode?: string;
    type: string;
    passengerType: string;
    ageRange?: string;
    ageRangeMin?: string;
    ageRangeMax?: string;
    proof?: string;
    proofDocuments?: string[];
    email: string;
    uuid: string;
    timeRestriction: FullTimeRestriction[];
    ticketPeriod: TicketPeriod;
    products: ProductDetails[];
    zoneName: string;
    stops: Stop[];
    additionalNocs: string[];
}

export const isSchemeOperatorTicket = (data: Ticket): data is SchemeOperatorTicket =>
    (data as SchemeOperatorTicket).schemeOperatorName !== undefined &&
    (data as SchemeOperatorTicket).schemeOperatorRegionCode !== undefined;

// Matching Data (created by the user on the site)

export interface PassengerDetails {
    passengerType: string;
    ageRangeMin?: string;
    ageRangeMax?: string;
    proofDocuments?: string[];
    email: string;
    uuid: string;
    timeRestriction: FullTimeRestriction[];
    ticketPeriod: TicketPeriodWithInput;
    proof?: string[];
}

export interface CompanionInfo {
    passengerType: string;
    minNumber?: string;
    maxNumber: string;
    ageRangeMin?: string;
    ageRangeMax?: string;
    proofDocuments?: string[];
}

export interface GroupPassengerInfo extends CompanionInfo {
    ageRange: string;
    proof: string;
}

export interface GroupDefinition {
    maxGroupSize: number;
    companions: CompanionInfo[];
}

export interface TimeRestriction {
    startTime?: string;
    endTime?: string;
    validDays: ('monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday' | 'sunday' | 'bankHoliday')[];
}

export interface ReturnPeriodValidity {
    amount: string;
    typeOfDuration: string;
}

export interface ReturnPeriodValidityWithErrors {
    amount?: string;
    typeOfDuration?: string;
    errors: ErrorInfo[];
}

export interface FareZone {
    name: string;
    stops: Stop[];
    prices: FareZonePrices[];
}

export interface FareZoneWithErrors {
    errors: ErrorInfo[];
}

export interface FareZonePrices {
    price: string;
    fareZones: string[];
}

export interface SelectedService {
    lineName: string;
    serviceCode: string;
    startDate: string;
    serviceDescription: string;
}

export interface BasicService {
    lineName: string;
    nocCode: string;
    operatorShortName: string;
    serviceDescription: string;
}

export interface SalesOfferPackageInfo {
    purchaseLocations: string[];
    paymentMethods: string[];
    ticketFormats: string[];
}

export interface SalesOfferPackageInfoWithErrors extends SalesOfferPackageInfo {
    errors: ErrorInfo[];
}

export interface SalesOfferPackage extends SalesOfferPackageInfo {
    name: string;
    description: string;
    id?: string;
}

export interface SalesOfferPackageWithErrors extends SalesOfferPackage {
    errors: ErrorInfo[];
}

export interface TicketPeriod {
    startDate?: string;
    endDate?: string;
}

export interface TicketPeriodWithInput extends TicketPeriod {
    dateInput: ProductDateInformation;
}

export interface BaseProduct {
    salesOfferPackages: SalesOfferPackage[];
}

export interface ProductWithSalesOfferPackages extends BaseProduct {
    productName: string;
}

export interface FlatFareProductDetails extends BaseProduct {
    productName: string;
    productPrice: string;
}

export interface ProductData {
    products: Product[];
}

export interface MultiOperatorInfoWithErrors {
    multiOperatorInfo: MultiOperatorInfo[];
    errors: ErrorInfo[];
}

export interface TimeBand {
    startTime: string;
    endTime: string;
}

export interface FullTimeRestriction {
    day: string;
    timeBands: TimeBand[];
}

export interface FullTimeRestrictionAttribute {
    fullTimeRestrictions: FullTimeRestriction[];
    errors: ErrorInfo[];
}

export interface TimeInput {
    timeInput: string;
    day: string;
}
export interface ProductDetails extends Product, BaseProduct {}
export interface FareStage {
    stageName: string;
    prices: {
        price: string;
        fareZones: string[];
    }[];
}

export interface UserFareStages {
    fareStages: FareStage[];
}

export interface UserFareZone {
    FareZoneName: string;
    NaptanCodes: string;
    AtcoCodes: string;
}

export interface DefinePassengerTypeWithErrors extends CompanionInfo {
    errors: ErrorInfo[];
}

export interface TimeRestrictionsDefinition extends TimeRestriction {
    timeRestrictionChoice?: string;
}

export interface TimeRestrictionsDefinitionWithErrors extends TimeRestrictionsDefinition {
    errors: ErrorInfo[];
}

export interface FareType {
    fareType: string;
}

export interface FareTypeWithErrors {
    errors: ErrorInfo[];
}

export interface GroupPassengerTypesCollection {
    passengerTypes: string[];
}
export interface GroupPassengerTypesCollectionWithErrors {
    errors: ErrorInfo[];
}

export interface MultiProduct {
    productName: string;
    productNameId: string;
    productPrice: string;
    productPriceId: string;
    productDuration: string;
    productDurationId: string;
    productDurationUnits: string;
    productDurationUnitsId: string;
    productValidity: string;
    productValidityId: string;
    productEndTime?: string;
    productEndTimeId?: string;
}

export interface MultiProductWithErrors extends MultiProduct {
    productNameError?: string;
    productPriceError?: string;
    productDurationError?: string;
    productDurationUnitsError?: string;
}

export interface PassengerType {
    passengerType: string;
    ageRange?: string;
    ageRangeMin?: string;
    ageRangeMax?: string;
    proof?: string;
    proofDocuments?: string[];
}

export interface PassengerTypeWithErrors {
    errors: ErrorInfo[];
}

export interface FaresInput {
    input: string;
    locator: string;
}

export interface FaresInformation {
    inputs: FaresInput[];
    errorInformation: PriceEntryError[];
}

export interface TicketPeriodWithErrors {
    errors: ErrorInfo[];
    dates: ProductDateInformation;
}

export interface ProductDateInformation {
    startDateDay: string;
    startDateMonth: string;
    startDateYear: string;
    endDateDay: string;
    endDateMonth: string;
    endDateYear: string;
}

export interface SelectSalesOfferPackageWithError {
    errors: ErrorInfo[];
    selected: { [key: string]: string[] };
}

export interface Service {
    service: string;
}

export interface ServiceWithErrors {
    errors: ErrorInfo[];
}

export interface ServiceType {
    lineName: string;
    startDate: string;
    description: string;
    origin?: string;
    destination?: string;
    serviceCode: string;
    dataSource?: string;
}

export interface ServicesInfo extends ServiceType {
    checked?: boolean;
}

export interface ServiceDB {
    serviceDescription: string;
    operatorShortName: string;
    journeyPatterns: JourneyPattern[];
}

export interface RawService {
    serviceDescription: string;
    operatorShortName: string;
    journeyPatterns: RawJourneyPattern[];
}

export interface JourneyPattern {
    startPoint: {
        Id: string;
        Display: string;
    };
    endPoint: {
        Id: string;
        Display: string;
    };
    stopList: string[];
}

export interface RawJourneyPattern {
    orderedStopPoints: {
        stopPointRef: string;
        commonName: string;
    }[];
}

// Components

export interface ConfirmationElement {
    name: string;
    content: string;
    href: string;
}

export interface FareTypeRadio {
    fareType: string;
    label: string;
}
export interface FareTypeRadioProps {
    standardFares: FareTypeRadio[];
    otherFares?: FareTypeRadio[];
}

export interface RadioWithoutConditionals extends BaseReactElement {
    value: string;
    radioButtonHint?: {
        id: string;
        content: string;
    };
}

export interface RadioWithConditionalInputs extends RadioWithoutConditionals {
    dataAriaControls: string;
    inputHint: {
        id: string;
        content: string;
        hidden?: boolean;
    };
    inputType: 'text' | 'checkbox' | 'date' | 'textWithUnits' | 'dropdown';
    inputs: BaseReactElement[] | PremadeTimeRestriction[];
    inputErrors: ErrorInfo[];
    selectIdentifier?: string;
}

export type RadioButton = RadioWithoutConditionals | RadioWithConditionalInputs;

export interface RadioConditionalInputFieldset {
    heading: {
        id: string;
        content: string;
        hidden?: boolean;
    };
    radios: RadioButton[];
    radioError: ErrorInfo[];
}

export interface UserAttribute {
    redirectFrom?: string;
}

export interface OperatorAttribute {
    name?: string;
    region?: string;
    nocCode?: string;
    uuid?: string;
    email?: string;
}

export interface ForgotPasswordAttribute {
    email: string;
}

export interface TxcSourceAttribute {
    source: 'tnds' | 'bods';
    hasTnds: boolean;
    hasBods: boolean;
}
