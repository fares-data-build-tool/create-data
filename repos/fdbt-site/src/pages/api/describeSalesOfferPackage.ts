import { NextApiResponse } from 'next';
import * as yup from 'yup';
import {
    defaultSalesOfferPackageOne,
    defaultSalesOfferPackageTwo,
    defaultSalesOfferPackageThree,
    defaultSalesOfferPackageFour,
} from '../selectSalesOfferPackage';
import { SOP_ATTRIBUTE, SOP_INFO_ATTRIBUTE } from '../../constants/attributes';
import { redirectToError, redirectTo, getAndValidateNoc } from './apiUtils';
import { NextApiRequestWithSession, ErrorInfo, SalesOfferPackage, SalesOfferPackageWithErrors } from '../../interfaces';
import { isSalesOfferPackageWithErrors } from '../describeSalesOfferPackage';
import { getSessionAttribute, updateSessionAttribute } from '../../utils/sessions';
import { isSalesOfferPackageInfoWithErrors } from '../salesOfferPackages';
import { insertSalesOfferPackage, getSalesOfferPackagesByNocCode } from '../../data/auroradb';

const noInputError = (input: string): string => `Enter a ${input} for your sales offer package`;
const inputTooLongError = (input: string, max: number): string => `Enter a ${input} that is ${max} characters or fewer`;

export const sopInfoSchema = yup.object({
    name: yup
        .string()
        .min(0, noInputError('name'))
        .max(20, inputTooLongError('name', 20))
        .required(noInputError('name')),
    description: yup
        .string()
        .min(0, noInputError('description'))
        .max(150, inputTooLongError('description', 150))
        .required(noInputError('description')),
});

export const checkAgainstDefaultNames = (sopName: string, errors: ErrorInfo[]): ErrorInfo[] => {
    if (
        sopName === defaultSalesOfferPackageOne.name ||
        sopName === defaultSalesOfferPackageTwo.name ||
        sopName === defaultSalesOfferPackageThree.name ||
        sopName === defaultSalesOfferPackageFour.name
    ) {
        errors.push({
            errorMessage: 'Sales offer package name cannot be the same as a default SOP name',
            id: 'sop-name',
            userInput: sopName,
        });
    }
    return errors;
};

export const checkUserInput = async (
    sopInfo: SalesOfferPackage,
): Promise<SalesOfferPackage | SalesOfferPackageWithErrors> => {
    let errors: ErrorInfo[] = [];
    try {
        await sopInfoSchema.validate(sopInfo, { abortEarly: false });
    } catch (validationErrors) {
        const validityErrors: yup.ValidationError = validationErrors;
        errors = validityErrors.inner.map(error => {
            return { errorMessage: error.message, id: `sop-${error.path}`, userInput: error.value };
        });
    }
    const updatedErrors = checkAgainstDefaultNames(sopInfo.name, errors);
    if (updatedErrors.length > 0) {
        return {
            ...sopInfo,
            errors: updatedErrors,
        };
    }
    return sopInfo;
};

export default async (req: NextApiRequestWithSession, res: NextApiResponse): Promise<void> => {
    try {
        const salesOfferPackageParams = getSessionAttribute(req, SOP_INFO_ATTRIBUTE);
        if (!salesOfferPackageParams || isSalesOfferPackageInfoWithErrors(salesOfferPackageParams)) {
            throw new Error('SOP_INFO_ATTRIBUTE is missing or in the wrong format');
        }

        const nocCode = getAndValidateNoc(req, res);
        if (!nocCode) {
            throw new Error('Could not retrieve nocCode from ID_TOKEN_COOKIE');
        }

        const { salesOfferPackageName, salesOfferPackageDescription } = req.body;
        let salesOfferPackageInfo: SalesOfferPackage = {
            name: salesOfferPackageName || '',
            description: salesOfferPackageDescription || '',
            purchaseLocations: salesOfferPackageParams.purchaseLocations,
            paymentMethods: salesOfferPackageParams.paymentMethods,
            ticketFormats: salesOfferPackageParams.ticketFormats,
        };

        salesOfferPackageInfo = await checkUserInput(salesOfferPackageInfo);

        if (isSalesOfferPackageWithErrors(salesOfferPackageInfo)) {
            updateSessionAttribute(req, SOP_ATTRIBUTE, salesOfferPackageInfo);
            redirectTo(res, '/describeSalesOfferPackage');
            return;
        }

        const savedSops = await getSalesOfferPackagesByNocCode(nocCode);
        const nameCheck = savedSops.find(sop => sop.name === salesOfferPackageInfo.name);

        if (nameCheck) {
            const sopWithError: SalesOfferPackageWithErrors = {
                ...salesOfferPackageInfo,
                errors: [
                    {
                        errorMessage: 'There is already a saved sales offer package with this name',
                        id: 'sop-name',
                        userInput: salesOfferPackageInfo.name,
                    },
                ],
            };
            updateSessionAttribute(req, SOP_ATTRIBUTE, sopWithError);
            redirectTo(res, '/describeSalesOfferPackage');
            return;
        }

        await insertSalesOfferPackage(nocCode, salesOfferPackageInfo);

        updateSessionAttribute(req, SOP_INFO_ATTRIBUTE, undefined);
        updateSessionAttribute(req, SOP_ATTRIBUTE, undefined);
        redirectTo(res, '/selectSalesOfferPackage');
    } catch (error) {
        const message = 'There was a problem on the describe sales offer package API.';
        redirectToError(res, message, 'api.describeSalesOfferPackage', error);
    }
};
