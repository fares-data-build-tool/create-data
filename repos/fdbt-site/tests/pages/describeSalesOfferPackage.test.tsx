import { shallow } from 'enzyme';
import React from 'react';
import DescribeSalesOfferPackage, { getServerSideProps } from '../../src/pages/describeSalesOfferPackage';
import { SalesOfferPackageInfo } from '../../src/pages/api/salesOfferPackages';
import { getMockContext } from '../testData/mockData';
import { SOP_INFO_ATTRIBUTE, SOP_ATTRIBUTE } from '../../src/constants';
import { SalesOfferPackage, SalesOfferPackageWithErrors } from '../../src/pages/api/describeSalesOfferPackage';
import { ErrorInfo } from '../../src/interfaces';

describe('describeSalesOfferPackage', () => {
    const mockError: ErrorInfo = expect.objectContaining({ errorMessage: expect.any(String), id: expect.any(String) });

    const mockSopInfoAttribute: SalesOfferPackageInfo = {
        purchaseLocations: ['OnBus', 'Shop', 'Mobile'],
        paymentMethods: ['Card', 'Cash'],
        ticketFormats: ['Paper', 'Mobile'],
    };
    const mockSopAttributeWithErrors: SalesOfferPackageWithErrors = {
        ...mockSopInfoAttribute,
        name: '',
        description: 'This is a sales offer package',
        errors: [mockError],
    };

    it('should render the page with no errors when no errors are present', () => {
        const tree = shallow(<DescribeSalesOfferPackage sopInfo={mockSopInfoAttribute} csrfToken="" />);
        expect(tree).toMatchSnapshot();
    });

    it('should render the page with errors when errors are present', () => {
        const tree = shallow(<DescribeSalesOfferPackage sopInfo={mockSopAttributeWithErrors} csrfToken="" />);
        expect(tree).toMatchSnapshot();
    });

    it('should be able to handle being passed undefined props', () => {
        const tree = shallow(<DescribeSalesOfferPackage sopInfo={undefined} csrfToken="" />);
        expect(tree).toMatchSnapshot();
    });

    describe('getServerSideProps', () => {
        it('should return props with no errors when only SOP_INFO_ATTRIBUTE is present', () => {
            const ctx = getMockContext({
                session: {
                    [SOP_INFO_ATTRIBUTE]: mockSopInfoAttribute,
                },
            });
            const expectedProps = {
                props: {
                    sopInfo: mockSopInfoAttribute,
                    csrfToken: '',
                },
            };
            const actualProps = getServerSideProps(ctx);
            expect(actualProps).toEqual(expectedProps);
        });
        it('should return props based on SOP_INFO_ATTRIBUTE when both SOP_INFO_ATTRIBUTE and SOP_ATTRIBUTE are present, but SOP_ATTRIBUTE contains no errors', () => {
            const sopAttribute: SalesOfferPackage = {
                ...mockSopInfoAttribute,
                name: 'Sales Offer Package',
                description: 'This is a sales offer package',
            };
            const ctx = getMockContext({
                session: {
                    [SOP_INFO_ATTRIBUTE]: mockSopInfoAttribute,
                    [SOP_ATTRIBUTE]: sopAttribute,
                },
            });
            const expectedProps = {
                props: {
                    sopInfo: mockSopInfoAttribute,
                    csrfToken: '',
                },
            };
            const actualProps = getServerSideProps(ctx);
            expect(actualProps).toEqual(expectedProps);
        });
        it('should return props with errors based on SOP_ATTRIBUTE when both SOP_INFO_ATTRIBUTE and SOP_ATTRIBUTE are present, but SOP_ATTRIBUTE contains errors', () => {
            const ctx = getMockContext({
                session: {
                    [SOP_INFO_ATTRIBUTE]: mockSopInfoAttribute,
                    [SOP_ATTRIBUTE]: mockSopAttributeWithErrors,
                },
            });
            const expectedProps = {
                props: {
                    sopInfo: mockSopAttributeWithErrors,
                    csrfToken: '',
                },
            };
            const actualProps = getServerSideProps(ctx);
            expect(actualProps).toEqual(expectedProps);
        });
    });
});
