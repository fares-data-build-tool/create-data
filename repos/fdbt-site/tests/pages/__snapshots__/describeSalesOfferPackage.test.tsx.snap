// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describeSalesOfferPackage should be able to handle being passed undefined props 1`] = `
<Component
  description="Sales Offer Package Description page of the Fares Data Build Tool"
  errors={Array []}
  title="Sales Offer Package Description - Fares Data Build Tool"
>
  <div
    className="govuk-grid-row"
  >
    <div
      className="govuk-grid-column-two-thirds"
    >
      <CsrfForm
        action="/api/describeSalesOfferPackage"
        csrfToken=""
        method="post"
      >
        <ErrorSummary
          errors={Array []}
        />
        <div
          className="govuk-form-group"
        >
          <fieldset
            aria-describedby="page-heading"
            className="govuk-fieldset"
          >
            <legend
              className="govuk-fieldset__legend govuk-fieldset__legend--l"
            >
              <h1
                className="govuk-fieldset__heading"
                id="page-heading"
              >
                Describe your sales offer package
              </h1>
            </legend>
            <span
              className="govuk-hint"
              id="describe-sop-hint"
            >
              Enter a name and description for your sales offer package that means something to you. This will be displayed back to you when you choose from your saved sales offer packages.
            </span>
            <div
              className="govuk-form-group"
            >
              <label
                className="govuk-label"
                htmlFor="sop-name"
              >
                Name
              </label>
              <FormElementWrapper
                errorClass="govuk-input--error"
                errorId=""
                errors={Array []}
              >
                <input
                  aria-describedby="describe-sop-hint"
                  className="govuk-input"
                  defaultValue=""
                  id="sop-name"
                  name="salesOfferPackageName"
                  type="text"
                />
              </FormElementWrapper>
            </div>
            <div
              className="govuk-form-group"
            >
              <label
                className="govuk-label"
                htmlFor="sop-description"
              >
                Description
              </label>
              <FormElementWrapper
                errorClass="govuk-textarea--error"
                errorId=""
                errors={Array []}
              >
                <textarea
                  aria-describedby="describe-sop-hint"
                  className="govuk-textarea"
                  id="sop-description"
                  name="salesOfferPackageDescription"
                  rows={4}
                />
              </FormElementWrapper>
            </div>
          </fieldset>
        </div>
        <input
          className="govuk-button"
          id="continue-button"
          type="submit"
          value="Continue"
        />
      </CsrfForm>
    </div>
    <div
      className="govuk-grid-column-one-third"
    >
      <Component />
    </div>
  </div>
</Component>
`;

exports[`describeSalesOfferPackage should render the page with errors when errors are present 1`] = `
<Component
  description="Sales Offer Package Description page of the Fares Data Build Tool"
  errors={
    Array [
      ObjectContaining {
        "errorMessage": Any<String>,
        "id": Any<String>,
      },
    ]
  }
  title="Sales Offer Package Description - Fares Data Build Tool"
>
  <div
    className="govuk-grid-row"
  >
    <div
      className="govuk-grid-column-two-thirds"
    >
      <CsrfForm
        action="/api/describeSalesOfferPackage"
        csrfToken=""
        method="post"
      >
        <ErrorSummary
          errors={
            Array [
              ObjectContaining {
                "errorMessage": Any<String>,
                "id": Any<String>,
              },
            ]
          }
        />
        <div
          className="govuk-form-group govuk-form-group--error"
        >
          <fieldset
            aria-describedby="page-heading"
            className="govuk-fieldset"
          >
            <legend
              className="govuk-fieldset__legend govuk-fieldset__legend--l"
            >
              <h1
                className="govuk-fieldset__heading"
                id="page-heading"
              >
                Describe your sales offer package
              </h1>
            </legend>
            <span
              className="govuk-hint"
              id="describe-sop-hint"
            >
              Enter a name and description for your sales offer package that means something to you. This will be displayed back to you when you choose from your saved sales offer packages.
            </span>
            <div
              className="govuk-form-group"
            >
              <label
                className="govuk-label"
                htmlFor="sop-name"
              >
                Name
              </label>
              <FormElementWrapper
                errorClass="govuk-input--error"
                errorId=""
                errors={
                  Array [
                    ObjectContaining {
                      "errorMessage": Any<String>,
                      "id": Any<String>,
                    },
                  ]
                }
              >
                <input
                  aria-describedby="describe-sop-hint"
                  className="govuk-input"
                  defaultValue=""
                  id="sop-name"
                  name="salesOfferPackageName"
                  type="text"
                />
              </FormElementWrapper>
            </div>
            <div
              className="govuk-form-group"
            >
              <label
                className="govuk-label"
                htmlFor="sop-description"
              >
                Description
              </label>
              <FormElementWrapper
                errorClass="govuk-textarea--error"
                errorId=""
                errors={
                  Array [
                    ObjectContaining {
                      "errorMessage": Any<String>,
                      "id": Any<String>,
                    },
                  ]
                }
              >
                <textarea
                  aria-describedby="describe-sop-hint"
                  className="govuk-textarea"
                  id="sop-description"
                  name="salesOfferPackageDescription"
                  rows={4}
                />
              </FormElementWrapper>
            </div>
          </fieldset>
        </div>
        <input
          className="govuk-button"
          id="continue-button"
          type="submit"
          value="Continue"
        />
      </CsrfForm>
    </div>
    <div
      className="govuk-grid-column-one-third"
    >
      <Component />
    </div>
  </div>
</Component>
`;

exports[`describeSalesOfferPackage should render the page with no errors when no errors are present 1`] = `
<Component
  description="Sales Offer Package Description page of the Fares Data Build Tool"
  errors={Array []}
  title="Sales Offer Package Description - Fares Data Build Tool"
>
  <div
    className="govuk-grid-row"
  >
    <div
      className="govuk-grid-column-two-thirds"
    >
      <CsrfForm
        action="/api/describeSalesOfferPackage"
        csrfToken=""
        method="post"
      >
        <ErrorSummary
          errors={Array []}
        />
        <div
          className="govuk-form-group"
        >
          <fieldset
            aria-describedby="page-heading"
            className="govuk-fieldset"
          >
            <legend
              className="govuk-fieldset__legend govuk-fieldset__legend--l"
            >
              <h1
                className="govuk-fieldset__heading"
                id="page-heading"
              >
                Describe your sales offer package
              </h1>
            </legend>
            <span
              className="govuk-hint"
              id="describe-sop-hint"
            >
              Enter a name and description for your sales offer package that means something to you. This will be displayed back to you when you choose from your saved sales offer packages.
            </span>
            <div
              className="govuk-form-group"
            >
              <label
                className="govuk-label"
                htmlFor="sop-name"
              >
                Name
              </label>
              <FormElementWrapper
                errorClass="govuk-input--error"
                errorId=""
                errors={Array []}
              >
                <input
                  aria-describedby="describe-sop-hint"
                  className="govuk-input"
                  defaultValue=""
                  id="sop-name"
                  name="salesOfferPackageName"
                  type="text"
                />
              </FormElementWrapper>
            </div>
            <div
              className="govuk-form-group"
            >
              <label
                className="govuk-label"
                htmlFor="sop-description"
              >
                Description
              </label>
              <FormElementWrapper
                errorClass="govuk-textarea--error"
                errorId=""
                errors={Array []}
              >
                <textarea
                  aria-describedby="describe-sop-hint"
                  className="govuk-textarea"
                  id="sop-description"
                  name="salesOfferPackageDescription"
                  rows={4}
                />
              </FormElementWrapper>
            </div>
          </fieldset>
        </div>
        <input
          className="govuk-button"
          id="continue-button"
          type="submit"
          value="Continue"
        />
      </CsrfForm>
    </div>
    <div
      className="govuk-grid-column-one-third"
    >
      <Component />
    </div>
  </div>
</Component>
`;
